### ISSUE TITLE
v2 lock-file format contains artifacts with missing dependencies

### ISSUE THREAD

**Issue by davidkurkov** [bazelbuild/rules_jvm_external#917]
We found that at least some artifacts defined in the v2 lock-file format are missing dependencies explicitly defined in an artifact's POM file which causes our builds/tests to fail.

For example, the [mockito-core 3.3.3 POM file](https://repo1.maven.org/maven2/org/mockito/mockito-core/3.3.3/mockito-core-3.3.3.pom) contains explicit dependencies on:
```xml
<dependencies>
    <dependency>
      <groupId>net.bytebuddy</groupId>
      <artifactId>byte-buddy</artifactId>
      <version>1.10.5</version>
      <scope>compile</scope>
    </dependency>
    <dependency>
      <groupId>net.bytebuddy</groupId>
      <artifactId>byte-buddy-agent</artifactId>
      <version>1.10.5</version>
      <scope>compile</scope>
    </dependency>
    <dependency>
      <groupId>org.objenesis</groupId>
      <artifactId>objenesis</artifactId>
      <version>2.6</version>
      <scope>compile</scope>
    </dependency>
  </dependencies>
```

In the v1-lockfile format, the `org.mockito:mockito-core:3.3.3` coordinate detailed the dependency on `bytebuddy`:

```json
{
    "coord": "org.mockito:mockito-core:3.3.3",
    "dependencies": [
        "net.bytebuddy:byte-buddy-agent:1.14.4",
        "net.bytebuddy:byte-buddy:1.14.4",
        "org.objenesis:objenesis:2.6"
    ],
    "directDependencies": [
        "net.bytebuddy:byte-buddy-agent:1.14.4",
        "net.bytebuddy:byte-buddy:1.14.4",
        "org.objenesis:objenesis:2.6"
    ],
}
```

In the v2-lockfile format, it is missing the `bytebuddy` dependency:

```json
{
  "artifacts": {
    "org.mockito:mockito-core": {
      "shasums": {
        "jar": "4be648c50456fba4686ba825000d628c1d805a3b92272ba9ad5b697dfa43036b",
        "sources": "b460650fbe05f5272e58df51628d3adff3dc870a3b19baf805bdea4d237aef1c"
      },
      "version": "3.3.3"
    }
  },
  "dependencies": {
    "org.mockito:mockito-core": [
      "org.objenesis:objenesis"
    ]
  },
  "skipped": [],
  "packages": {
    "org.mockito:mockito-core": []
  },
  "repositories": {
    "https://maven.google.com/": [
      "org.mockito:mockito-core",
      "org.mockito:mockito-core:jar:sources"
    ]
  },
  "version": "2"
}
```

@artem-zinnatullin and I are looking into what in the changes for the v2 lockfile could be causing this but wanted to raise an issue in case anyone else has come across this or might have more context.

**Comment by rdesgroppes** [comment]
I was not able to reproduce the issue (missing deps) on top of above mentioned commits, see: https://github.com/bazelbuild/rules_jvm_external/pull/918#issuecomment-1597372594
If you could share your exact `maven_install`, I'd be glad to have a further look.

**Comment by artem-zinnatullin** [comment]
Saw failed tests on my PR #918 and comments, trying to reproduce the issue we're seeing on a smaller installation and see what else can cause it in our case, sorry for potentially bad report 😰

**Comment by artem-zinnatullin** [comment]
@rdesgroppes @cheister

1. My sincere apologies for submitting bad PR #918, closed it
2. I was able to trace down our 470 dependencies graph to the cause of the issue and filed a PR with minimalistic regression test that you can run on your end #919

TL;TR:

1. `mockito-core` depends on `byte-buddy`
2. `androidx.arch.core:core-testing` [POM](https://dl.google.com/android/maven2/androidx/arch/core/core-testing/2.1.0/core-testing-2.1.0.pom) has `exclusion` for `<groupId>net.bytebuddy</groupId>`
3. If both are present in `maven_install` then `mockito-core` doesn't get its `byte-buddy` dependency
4. If `androidx.arch.core:core-testing` is removed from `maven_install` then `mockito-core` gets `byte-buddy` dependency
5. Adding `byte-buddy` manually to `maven_install` doesn't help
6. `mockito-core` was getting `byte-buddy` dependency with `androidx.arch.core:core-testing` in `maven_install` prior to change #908 
7. This is likely a bug in coursier due to how #908 passes classifier and type for dependency resolution
8. `byte-buddy` also doesn't list its `<packaging>jar</packaging>` in its [POM](https://repo1.maven.org/maven2/net/bytebuddy/byte-buddy/1.14.4/byte-buddy-1.14.4.pom), but I'm not sure that's related

**Comment by rdesgroppes** [comment]
> 8. `byte-buddy` also doesn't list its `<packaging>jar</packaging>` in its [POM](https://repo1.maven.org/maven2/net/bytebuddy/byte-buddy/1.14.4/byte-buddy-1.14.4.pom), but I'm not sure that's related

This is the default as per [POM Reference](https://maven.apache.org/pom.html#packaging):
> When no packaging is declared, Maven assumes the packaging is the default: `jar`. [...]

**Comment by rdesgroppes** [comment]
> 1. My sincere apologies for submitting bad PR [...]

Quite the opposite: thanks for surfacing a bug!

**Comment by rdesgroppes** [comment]
Setting the `RJE_VERBOSE=1` environment variable prior to pinning indeed reveals a wrong item being passed to Coursier:
```sh
argsfile_content:
"-noverify"
"-jar"
"/path/to/external/unpinned_maven/coursier"
fetch
[...]
"androidx.arch.core:core-testing:aar,type=2.1.0"
```
With `maven_install(version_conflict_policy="pinned")`, one additionally gets:
```
"--force-version"
"androidx.arch.core:core-testing:aar"
```

**Comment by artem-zinnatullin** [comment]
@rdesgroppes thanks for finding a simple clever fix, updated PR #919 with it, should be ready for merge!

**Comment by rdesgroppes** [comment]
Let's find a benevolent reviewer :)
